File: src/binary/Test1.java
package binary;

import java.util.Arrays;

public class Test1  {
    public static void main(String[] args) {
//        int[] budgets = {120, 110, 140, 150};
//        int M = 485;

        int[] budgets = {5, 5, 5, 5, 15, 30};
        int M = 62;
        System.out.println(solution(budgets, M));
    }

    static int solution(int[] budgets, int M) {
        Arrays.sort(budgets);
        int answer = 0;
        if (budgets[0] >= M) {
            return M / budgets.length;
        }
        for (int i = 0; i < budgets.length; i++) {
            M -= budgets[i];
            if (answer > M) {
                break;
            }
            answer = M / (budgets.length - i - 1);
        }
        if (budgets[budgets.length - 1] <= answer) {
            answer = budgets[budgets.length - 1];
        }
        return answer;
    }
}

File: src/sort/Test1.kt
package sort

import java.util.*

fun main() {
    val array = intArrayOf(1, 5, 2, 6, 3, 7, 4)
    val commands = arrayOf(intArrayOf(2, 5, 3), intArrayOf(4, 4, 1), intArrayOf(1, 7, 3))
    solution(array, commands).forEach {
        print(it)
    }
}

private fun solution(array: IntArray, commands: Array<IntArray>): IntArray {
    val answer = IntArray(commands.size)
    commands.forEachIndexed { i, ints ->
        val temp = Arrays.copyOfRange(array, ints[0] - 1, ints[1])
        Arrays.sort(temp)
        answer[i] = temp[ints[2] - 1]
    }
    return answer
}